akka-actor-typed/src/main/scala/akka/actor/typed/internal/StashBufferImpl.scala

```diff
@@ -6,150 +6,80 @@ package akka.actor.typed.internal
-
-  override def isFull: Boolean = size == capacity
-
-  override def stash(message: T): ImmutableStashBufferImpl[T] = {
-    if (message == null) throw new NullPointerException
-    if (isFull)
-      throw new javadsl.StashOverflowException(s"Couldn't add [${message.getClass.getName}] " +
-        s"because stash with capacity [$capacity] is full")
-
-    new ImmutableStashBufferImpl(capacity, buffer :+ message)
-  }
-
-  override def dropHead(): ImmutableStashBufferImpl[T] =
-    if (buffer.nonEmpty) new ImmutableStashBufferImpl(capacity, buffer.tail)
-    else throw new NoSuchElementException("head of empty buffer")
-
-  override def drop(numberOfMessages: Int): ImmutableStashBufferImpl[T] =
-    if (isEmpty) this
-    else new ImmutableStashBufferImpl(capacity, buffer.drop(numberOfMessages))
-
-  override def head: T =
-    if (buffer.nonEmpty) buffer.head
-    else throw new NoSuchElementException("head of empty buffer")
-
-  override def foreach(f: T ⇒ Unit): Unit =
-    buffer.foreach(f)
-
-  override def forEach(f: Consumer[T]): Unit = foreach(f.accept)
-
-  override def unstashAll(ctx: scaladsl.ActorContext[T], behavior: Behavior[T]): Behavior[T] =
-    unstash(ctx, behavior, size, ConstantFun.scalaIdentityFunction[T])
-
-  override def unstashAll(ctx: javadsl.ActorContext[T], behavior: Behavior[T]): Behavior[T] =
-    unstashAll(ctx.asScala, behavior)
-
-  override def unstash(scaladslCtx: scaladsl.ActorContext[T], behavior: Behavior[T],
-                       numberOfMessages: Int, wrap: T ⇒ T): Behavior[T] = {
-    val ctx = scaladslCtx.asInstanceOf[ActorContext[T]]
-    val iter = buffer.iterator.take(numberOfMessages).map(wrap)
-    Behavior.interpretMessages[T](behavior, ctx, iter)
-  }
-
-  override def unstash(ctx: javadsl.ActorContext[T], behavior: Behavior[T],
-                       numberOfMessages: Int, wrap: JFunction[T, T]): Behavior[T] =
-    unstash(ctx.asScala, behavior, numberOfMessages, x ⇒ wrap.apply(x))
-
-}
-
-/**
- * INTERNAL API
- */
-@InternalApi private[akka] object MutableStashBufferImpl {
+@InternalApi private[akka] object StashBufferImpl {
   private final class Node[T](var next: Node[T], val message: T) {
     def apply(f: T ⇒ Unit): Unit = f(message)
   }
 
-  def apply[T](capacity: Int): MutableStashBufferImpl[T] =
-    new MutableStashBufferImpl(capacity, null, null)
+  def apply[T](capacity: Int): StashBufferImpl[T] =
+    new StashBufferImpl(capacity, null, null)
 }
 
 /**
  * INTERNAL API
  */
-@InternalApi private[akka] final class MutableStashBufferImpl[T] private (
-  val capacity:      Int,
-  private var _head: MutableStashBufferImpl.Node[T],
-  private var _tail: MutableStashBufferImpl.Node[T])
-  extends javadsl.MutableStashBuffer[T] with scaladsl.MutableStashBuffer[T] {
+@InternalApi private[akka] final class StashBufferImpl[T] private (
+  val capacity:       Int,
+  private var _first: StashBufferImpl.Node[T],
+  private var _last:  StashBufferImpl.Node[T])
```

Comment:
@chbatey you had a question if tail was the last. I renamed them to first and last, which answers your question.